apps/librelingo_audios/librelingo_audios/generate_audios.py

```diff
@@ -0,0 +1,88 @@
+from librelingo_utils import audio_id, iterate_phrases
+from pathlib import Path
+import subprocess
+import json
+import random
+
+def generate_audios_for_course(output_path, course_name, course, settings):
+    if not course.settings.audio_settings.enabled:
+        return
+
+    index_path = Path(Path(output_path) / "{}.json".format(course_name))
+
+    phrases_in_course = [(_index_dict_for_phrase(p, course), p) for p in iterate_phrases(course)]
+
+    if settings.overwrite:
+        # Pretend nothing already exists; recreate everything
+        phrase_indices_already_present = []
+    else:
+        with open(index_path, "r") as f:
+            phrase_indices_already_present = json.loads(f.read())["audios"]
+
+    result_index = []
+
+    # For each phrase in the course, if it doesn't exist, generate it.
+    for course_phrase_index, course_phrase in phrases_in_course:
+        # TODO check for equality along multiple dimensions when there are
+        # multiple sources
+        existing_matching_index = next((i for i in phrase_indices_already_present if i["id"] == course_phrase_index["id"]), None)
+
+        if existing_matching_index is None:
+            _generate_audio_for_phrase(course_phrase, output_path, course, settings)
+            result_index.append(course_phrase_index)
+        else:
+            result_index.append(existing_matching_index)
+
+    # For phrases that already exist as audio, check if they are still needed by
+    # the course.
+    for index in phrase_indices_already_present:
+        matching_course_phrase = next(((i, p) for i, p in phrases_in_course if i["id"] == index["id"]), None)
+
+        if matching_course_phrase is None:
+            _delete_audio_for_phrase(course_phrase, output_path, course, settings)
+
+    if not settings.dry_run:
+        with open(index_path, "w", encoding="utf-8") as f:
+            json.dump({"audios": result_index}, f, ensure_ascii=False, indent=4)
```

Comment:
Why not?

```suggestion
            json.dump(result_index, f, ensure_ascii=False, indent=4)
```

Reply:
Also why ensure_ascii=False?

Reply:
I think just JSON-serializing the array is fine, not sure why I went with a wrapper object. `ensure_ascii=False` allows Unicode characters in the file.

Reply:
I am not sure what the advantages of unicode vs ascii escape sequences are? 🤔 I guess both should work well so I don't mind either way